File: src/main/js/monster.ts
interface Monster {

    id: number;
    name: string;
    alignment: Alignment[];
    environments: string[];
    size: Size;
    type: Type;
    bookSource: BookSource;
    armourClass: ArmourClass[];
    hp: HP;
    speed: Speed;

    // Attributes
    familiar: number;
    strength: number;
    dexterity: number;
    constitution: number;
    intelligence: number;
    wisdom: number;
    charisma: number;

    //saves
    strengthSave: number;
    dexteritySave: number;
    constitutionSave: number;
    intelligenceSave: number;
    wisdomSave: number;
    charismaSave: number;

    //skills
    athletics: number;
    acrobatics: number;
    slightOfHand: number;
    arcana: number;
    history: number;
    investigation: number;
    nature: number;
    religion: number;
    animalHandling: number;
    insight: number;
    medicine: number;
    perception: number;
    survival: number;
    deception: number;
    intimidation: number;
    performance: number;
    persuasion: number;
    passivePerception: number;
    stealth: number;

    languages: string[];
    challengeRating: ChallengeRating;

    senses: string[];
    trait: Action[];
    action?: Action[];
    legendaryAction?: Action[];
    reaction?: Action[];

    immunities: Immunities[];
    vulnerabilities: Immunities[];
    resistances: Resistances[];
    conditionImmunity: ConditionImmunity[];
}

enum Size {
    TINY = "TINY",
    SMALL = "SMALL",
    MEDIUM = "MEDIUM",
    LARGE = "LARGE",
    HUGE = "HUGE",
    GARGANTUAN = "GARGANTUAN",
}

interface Alignment {
    alignemnt: string;
    chance: number;
}

interface Type {
    type: string;
    subtype: string[];
    swarmSize: number;
}

interface BookSource {
    bookCode: string;
    book: string;
    page: number;
}

interface ArmourClass {
    armourClass: number;
    armourSources: string;
    condition: string;
}

interface HP {
    averageHp: number;
    numOfDice: number;
    sizeOfDie: number;
    flatBonus: number;
    special: number;
}

interface Speed {
    walk: number;
    walkCondition: string;
    burrow: number;
    burrowCondition: string;
    climb: number;
    climbCondition: string;
    fly: number;
    flyCondition: string;
    swim: number;
    swimCondition: string;
    hover: boolean;
}

/* Valid settings
0 1/8 1/4 1/2 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 26 30
*/
interface ChallengeRating {
    challengeRating: number;
    coven: string;
    lair: string;
}

interface Entry {
    name: string;
    entries: string[];
    attack: null;
    subEntries: string[];
}

interface Action {
    name: string;
    entries: string[];
    attack?: number;
    subEntries?: Entry[];
}

interface Immunities {
    damageType: string;
    condition: string;
}

interface Resistances {
    damageType: string;
    note: string;
    preNote: null;
    special: null;
}

interface ConditionImmunity {
    condition: string;
    note: string;
}

File: src/main/java/com/dnd/tools/encounterhelper/monster/model/Monster.java
package com.dnd.tools.encounterhelper.monster.model;

import java.util.List;
import java.util.Set;
import javax.persistence.CascadeType;
import javax.persistence.ElementCollection;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import lombok.Data;

@Data
@Entity
public class Monster {

  @Id
  @GeneratedValue
  private Long id;

  private String name;
  @ElementCollection
  private List<AlignmentOption> alignment;
  @ElementCollection
  private Set<Environment> environments;
  @Enumerated(EnumType.STRING)
  private Size size;
  @Embedded
  private Type type;
  @Embedded
  private BookSource bookSource;
  @ElementCollection
  private List<ArmourClass> armourClass;
  @Embedded
  private Hp hp;
  @Embedded
  private Speed speed;
  private Boolean familiar;
  
  private Integer strength;
  private Integer dexterity;
  private Integer constitution;
  private Integer intelligence;
  private Integer wisdom;
  private Integer charisma;

  private Integer strengthSave;
  private Integer dexteritySave;
  private Integer constitutionSave;
  private Integer intelligenceSave;
  private Integer wisdomSave;
  private Integer charismaSave;

  private Integer athletics;
  private Integer acrobatics;
  private Integer slightOfHand;
  private Integer Stealth;
  private Integer arcana;
  private Integer history;
  private Integer investigation;
  private Integer nature;
  private Integer religion;
  private Integer animalHandling;
  private Integer insight;
  private Integer medicine;
  private Integer perception;
  private Integer survival;
  private Integer deception;
  private Integer intimidation;
  private Integer performance;
  private Integer persuasion;

  @ElementCollection
  private List<Resistance> resistances;
  @ElementCollection
  private List<Immunity> immunities;
  @ElementCollection
  private List<Vulnerability> vulnerabilities;
  @ElementCollection
  private Set<ConditionImmunity> conditionImmunity;

  @ElementCollection
  private Set<String> senses;
  private Integer passivePerception;
  @ElementCollection
  private Set<String> languages;
  @Embedded
  private ChallengeRating challengeRating;

  @OneToMany(cascade = CascadeType.PERSIST)
  private List<Ability> trait;
  @OneToMany(cascade = CascadeType.PERSIST)
  private List<Ability> action;
  @OneToMany(cascade = CascadeType.PERSIST)
  private List<Ability> reaction;
  @OneToMany(cascade = CascadeType.PERSIST)
  private List<Ability> legendaryAction;

  @OneToMany(cascade = CascadeType.PERSIST)
  private List<Variant> variants;

  @Embedded
  private InnateSpellCasting innateSpellCasting;
  @Embedded
  private Spellcasting spellcasting;

  @OneToMany(cascade = CascadeType.PERSIST)
  private List<LairAction> lairActions;
  @OneToMany(cascade = CascadeType.PERSIST)
  private List<RegionalEffect> regionalEffects;
}
